src/Bicep.Core/Syntax/JsonSyntaxTree.cs

```diff
@@ -0,0 +1,20 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Immutable;
+using Newtonsoft.Json.Linq;
+
+namespace Bicep.Core.Syntax
+{
+    public class JsonSyntaxTree : SyntaxTree
```

Comment:
Have you considered representing the template semantically without actually generating a syntax tree? My concern with extending `SyntaxTree` is just that it's a corner case to maintain with quite different behavior. e.g. when raising diagnostics or doing type analysis, it'll be a subtlety that could easily be missed, or add complexity to the code, when it feels like there are real meaningful differences between a 'real' syntax tree and a generated one. Providing an accurate decompilation also feels like it could be a maintenance overhead in the long run which may not be necessary if we avoid generating syntax.

Reply:
Yes, actually my original design was like that (to create a different semantic module for ARM templates), but I switch to the current approach to the existing semantic model and type checker as much as possible. Now I realize it might not be worth it given the maintenance overhead. @majastrz also raised the a similar concern (in the comments on adding tests to ensure the converters and the decompiler logic are in sync). It won't be too hard to switch back to my original design since I still have the code, but I would like to have a discussion with you and Marcin before moving forward.

Reply:
As we discussed offline, I've refactored the code to create a different semantic model for JSON templates, which removes the converters and avoids extending `SyntaxTree`.